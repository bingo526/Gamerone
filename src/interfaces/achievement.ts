/**
 * Gamer One API Documentation
 * Welcome to the Official Gamer One API documentation.
 *
 * OpenAPI spec version: 1.0.0
 * Contact: developer@gamerone.gg
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 * Do not edit the class manually.
 */
import { Badge } from './badge';

export type ActionType =
  | ''
  | 'up-vote'
  | 'comment'
  | 'experience'
  | 'follow'
  | 'follower'
  | 'post'
  | 'profile_complete'
  | 'signup'
  | 'game'
  | 'currently_playing'
  | 'sponsor'
  | 'socialNetwork'
  | 'store'
  | 'gear';

export interface Achievement {
  id: number;
  date: Date;
  badge: Badge | null;
  action: ActionType;
  threshold: number;
}
