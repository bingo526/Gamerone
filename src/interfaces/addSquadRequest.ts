/**
 * Gamer One API Documentation
 * Welcome to the Official Gamer One API documentation.
 *
 * OpenAPI spec version: 1.0.0
 * Contact: developer@gamerone.gg
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 * Do not edit the class manually.
 */

export interface AddSquadRequest {
  /**
   * User id for the user that you want to add to your squad.
   */
  friendId: number;
}
