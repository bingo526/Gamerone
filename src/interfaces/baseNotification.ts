/**
 * Gamer One API Documentation
 * Welcome to the Official Gamer One API documentation.
 *
 * OpenAPI spec version: 1.0.0
 * Contact: developer@g1.gg
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 * Do not edit the class manually.
 */

export interface BaseNotification {
  /**
   * UUID for the notification ID
   */
  id: string;
  /**
   * Notification generater. Identifies what type of notification this is
   */
  type: NoficationType;
  /**
   * Indicates when the notification is marked as read. Null if the notification is unread.
   */
  readAt: Date;
  /**
   * Indicates when the notification created
   */
  createdAt: Date;
}

export type NoficationType =
  | 'G1\\Notifications\\PostMention'
  | 'G1\\Notifications\\SquadRequest'
  | 'G1\\Notifications\\SquadRequestAccepted';

export enum NoficationTypeEnum {
  PostMention = 'G1\\Notifications\\PostMention',
  SquadRequest = 'G1\\Notifications\\SquadRequest',
  SquadRequestAccepted = 'G1\\Notifications\\SquadRequestAccepted',
}
