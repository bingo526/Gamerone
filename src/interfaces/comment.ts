/**
 * Gamer One API Documentation
 * Welcome to the Official Gamer One API documentation.
 *
 * OpenAPI spec version: 1.0.0
 * Contact: developer@gamerone.gg
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 * Do not edit the class manually.
 */
import { User } from './user';

export interface Comment {
  id: number;
  comment: string;
  createdAt: string;
  user: User;
}
