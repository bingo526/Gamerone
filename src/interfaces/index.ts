export * from './common';
export * from './achievement';
export * from './addSquadRequest';
export * from './badge';
export * from './baseNotification';
export * from './checkEmailRequest';
export * from './checkUsernameRequest';
export * from './club';
export * from './clubPermissionLevelRequest';
export * from './clubRequest';
export * from './clubType';
export * from './comment';
export * from './commentPagedResponse';
export * from './nowPlaying';
export * from './nowPlayingRequest';
export * from './event';
export * from './eventRequest';
export * from './experienceType';
export * from './forgotPasswordRequest';
export * from './game';
export * from './gameAddRequest';
export * from './gameCategory';
export * from './gamePlatform';
export * from './gameUpdateRequest';
export * from './gear';
export * from './gearType';
export * from './inlineResponse200';
export * from './layoutSettings';
export * from './loginRequest';
export * from './loginResponse';
export * from './modelError';
export * from './multiFollowRequest';
export * from './notification';
export * from './notifications';
export * from './pagedPost';
export * from './pagedResponse';
export * from './post';
export * from './postCommentRequest';
export * from './postMedia';
export * from './postMentionNotification';
export * from './postMentionNotificationData';
export * from './postRequest';
export * from './product';
export * from './productRequest';
export * from './profile';
export * from './profileSettingsRequest';
export * from './resetPasswordRequest';
export * from './route';
export * from './signupRequest';
export * from './socialNetwork';
export * from './socialNetworkRequest';
export * from './socialNetworkType';
export * from './sponsor';
export * from './sponsorRequest';
export * from './squadInviteList';
export * from './squadList';
export * from './squadRequestNotification';
export * from './squadRequestNotificationData';
export * from './squadRequestAcceptedNotification';
export * from './squadRequestAcceptedNotificationData';
export * from './squadOutgoingList';
export * from './statusRequest';
export * from './unfurlRequest';
export * from './unfurlURL';
export * from './uploadImageRequest';
export * from './uploadRequest';
export * from './user';
export * from './userExperience';
export * from './userExperienceAddRequest';
export * from './userGame';
export * from './userGamePagedResponse';
export * from './userGear';
export * from './userGearAddRequest';
export * from './userGearPagedResponse';
export * from './userPagedResponse';
export * from './userPrivacy';
export * from './validationError';
export * from './verifyTokenRequest';
export * from './widgetReferenceData';
