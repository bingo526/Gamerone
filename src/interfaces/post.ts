/**
 * Gamer One API Documentation
 * Welcome to the Official Gamer One API documentation.
 *
 * OpenAPI spec version: 1.0.0
 * Contact: developer@g1.gg
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 * Do not edit the class manually.
 */
import { Game } from './game';
import { PostMedia } from './postMedia';

export interface Post {
  id: number;
  title: string;
  description: string | null;
  image: string;
  url: string;
  html: string;
  privacy: string;
  commentsEnabled: boolean;
  isVoted: PostIsVoted;
  upCount: number;
  downCount: number;
  commentCount: number;
  createdAt: string;
  games: Array<Game>;
  images: Array<PostMedia>;
}
export type PostIsVoted = 'null' | 'up' | 'down';

export enum PostIsVotedEnum {
  Null = 'null',
  Up = 'up',
  Down = 'down',
}
