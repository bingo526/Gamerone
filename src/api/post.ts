import { PostCommentRequest } from './../interfaces/postCommentRequest';
import request, { stringifyBody } from 'utils/request';
import { PostRequest, Post, StatusResponse, PagedResponse } from 'interfaces';

const requestOptions = (param: any, method = 'POST') => ({
  method,
  body: stringifyBody(param),
});

export interface GetPostsParam {
  userId: number;
}

export interface GetPostCommentsParam {
  postId: number;
}

/**
 * Add a new post
 */
export const addNewPost = (param: PostRequest) => {
  return request<Post>(`/posts`, requestOptions(param));
};

/**
 * Load user's feed posts
 */
export const loadFeed = () => {
  return request<Post[]>('/posts');
};

/**
 * Get profile's posts
 */
export const getPosts = (param: GetPostsParam) => {
  return request<Post[]>(`/posts/user/${param.userId}`);
};

/**
 * Get a post
 */
export const getPost = (id: number) => {
  return request<Post>(`/posts/${id}`);
};

/**
 * Update a post
 */
export const updatePost = (id: number, post: Post) => {
  return request<Post>(`/posts/${id}`, requestOptions(post, 'PATCH'));
};

/**
 * Delete a post
 */
export const deletePost = (id: number) => {
  return request<StatusResponse>(`/posts/${id}`, {
    method: 'DELETE',
  });
};

/**
 * Up vote post
 */
export const upVotePost = (id: number) => {
  return request<Post>(`/posts/${id}/up`);
};

/**
 * Down vote post
 */
export const downVotePost = (id: number) => {
  return request<Post>(`/posts/${id}/down`);
};

/*
 * Add a new comment to the post
 */
export const addComment = (postId: number, param: PostCommentRequest) => {
  return request<Comment>(
    `/posts/${postId}/comments`,
    requestOptions(param, 'POST'),
  );
};

/**
 * List comments related to the post.
 */
export const getPostComments = (param: GetPostCommentsParam, page = 0) => {
  return request<PagedResponse>(`/posts/${param.postId}/comments?page=${page}`);
};

/**
 * Get the comment with given id
 */
export const getComment = (commentId: number) => {
  return request<Comment>(`/posts/comments/${commentId}`);
};

/**
 * Delete the comment with given id
 */
export const deleteComemnt = (commentId: number) => {
  return request<StatusResponse>(
    `/posts/comments/${commentId}`,
    requestOptions({}, 'DELETE'),
  );
};
