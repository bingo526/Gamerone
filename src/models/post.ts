import {
  Post,
  Game,
  PostMedia,
  PostIsVoted,
  PostIsVotedEnum,
} from 'interfaces';
import { UserModel } from 'models/user';
import { POST_PLACEHOLDER } from 'utils/constants';
import { PostUser } from 'interfaces/postUser';

// TODO: Set correct default data
export class PostModel implements Post {
  id = 0;
  title = '';
  description: string | null = null;
  image = POST_PLACEHOLDER;
  url = '';
  html = '';
  privacy = '';
  commentsEnabled = true;
  isVoted: PostIsVoted = PostIsVotedEnum.Null;
  upCount = 0;
  downCount = 0;
  commentCount = 0;
  createdDate = new Date();
  user = new UserModel();
  games: Game[] = [];
  images: PostMedia[] = [];

  // TODO: post can be PostUser or PostClub
  fromDto = (post: Post) => {
    this.id = post.id;
    this.title = post.title;
    this.description = post.description;
    this.image = post.image;
    this.url = post.url;
    this.html = post.html;
    this.privacy = post.privacy;
    this.commentsEnabled = post.commentsEnabled;
    this.isVoted = post.isVoted;
    this.upCount = post.upCount;
    this.downCount = post.downCount;
    this.commentCount = post.commentCount;
    this.createdDate = new Date(post.createdAt);
    this.games = post.games.map((game) => game);
    this.images = post.images.map((image) => image);
    return this;
  };

  fromPostUserDto = (post: PostUser) => {
    this.fromDto(post);
    this.user = new UserModel().fromDto(post.user);
    return this;
  };

  get createdAt() {
    return this.createdDate.toISOString();
  }
}
