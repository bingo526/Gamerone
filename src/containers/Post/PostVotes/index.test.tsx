import React from 'react';
import { render, RenderResult, fireEvent } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import { PostVotes } from './index';
import { MappedProps as PostVotesStoreProps } from './index';
import { PostModel } from 'models/post';
import { PostIsVotedEnum } from 'interfaces';

const post = new PostModel().fromDto({
  id: 20,
  title: 'First Post',
  description: 'This is my first post!!!',
  privacy: 'public',
  commentsEnabled: true,
  html: '',
  url: '',
  image: 'assets/placeholder/post.svg',
  createdAt: '2020-05-22T21:57:57.000000Z',
  upCount: 5,
  downCount: 3,
  commentCount: 0,
  user: {
    id: 1481,
    username: 'specialk',
    email: null,
    firstName: 'Kevin',
    lastName: 'Manship',
    websiteUrl: 'https://www.g1.gg',
    bio:
      'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc scelerisque lacus quis euismod ultricies. Ut ullamcorper porta eros vel mollis. Praesent non est nulla. Curabitur sit amet convallis odio. Praesent et augue erat.',
    status: null,
    location: 'Vancouver, CA',
    avatar:
      'user_content/1481/avatar/rKURZsZcbXHrfPtktVnSD0SM147bgyokvfqoI9GZ.png',
    banner:
      'user_content/1481/banner/FwWhmQnzxZrYlxndnGKtbvs7BoPfAVlNbZg8IYoQ.png',
    playCount: 0,
    gameCount: 0,
    badgeCount: 0,
    followCount: 1,
    followerCount: 5,
    birthDate: null,
    isFollowing: false,
  },
  images: [
    {
      filename:
        'user_content/1481/post/P64y8PuF5hqQT2QfOttqhDV8VHAR1583M1GZSCog.jpeg',
    },
    {
      filename:
        'user_content/1481/post/tiKMdzxeH5jrY4Fk9JUPj1x5mpn6r1ZyVHADPq2T.jpeg',
    },
  ],
  games: [],
  isVoted: null,
});

const renderComponent = ({
  isAuthenticated,
  dispatchUpVote,
  dispatchDownVote,
}: PostVotesStoreProps): RenderResult | undefined => {
  let utils;

  act(() => {
    utils = render(
      <PostVotes
        post={post}
        isAuthenticated={isAuthenticated}
        dispatchUpVote={dispatchUpVote}
        dispatchDownVote={dispatchDownVote}
      />,
    );
  });

  return utils;
};

test('renders correct number counts', () => {
  const dispatchUpVote = jest.fn();
  const dispatchDownVote = jest.fn();
  const { asFragment, container } = renderComponent({
    isAuthenticated: true,
    dispatchUpVote,
    dispatchDownVote,
  }) as RenderResult;

  expect(asFragment()).toMatchSnapshot();

  expect(container.querySelector('#up-vote-count')?.textContent).toEqual(
    post.upCount.toString(),
  );

  expect(container.querySelector('#down-vote-count')?.textContent).toEqual(
    post.downCount.toString(),
  );

  expect(container.querySelector('#comment-count')?.textContent).toEqual(
    post.commentCount.toString(),
  );
});

test('up/vote actions should be disabled when not authenticated', () => {
  const dispatchUpVote = jest.fn();
  const dispatchDownVote = jest.fn();
  const { getByLabelText } = renderComponent({
    isAuthenticated: false,
    dispatchUpVote,
    dispatchDownVote,
  }) as RenderResult;

  act(() => {
    fireEvent.click(getByLabelText('up-vote'));
  });

  expect(dispatchUpVote).toHaveBeenCalledTimes(0);

  act(() => {
    fireEvent.click(getByLabelText('down-vote'));
  });

  expect(dispatchDownVote).toHaveBeenCalledTimes(0);
});

test('shows up voted status when isVoted is up', () => {
  const dispatchUpVote = jest.fn();
  const dispatchDownVote = jest.fn();

  // set isVoted up
  post.isVoted = PostIsVotedEnum.Up;

  const { getByLabelText } = renderComponent({
    isAuthenticated: false,
    dispatchUpVote,
    dispatchDownVote,
  }) as RenderResult;

  expect(
    getByLabelText('up-vote').classList.contains('button--primary'),
  ).toBeTruthy();
});

test('shows down voted status when isVoted is down', () => {
  const dispatchUpVote = jest.fn();
  const dispatchDownVote = jest.fn();

  // set isVoted down
  post.isVoted = PostIsVotedEnum.Down;

  const { getByLabelText } = renderComponent({
    isAuthenticated: false,
    dispatchUpVote,
    dispatchDownVote,
  }) as RenderResult;

  expect(
    getByLabelText('down-vote').classList.contains('button--primary'),
  ).toBeTruthy();
});

test('increases up voted count when up vote', () => {
  const dispatchUpVote = jest.fn();
  const dispatchDownVote = jest.fn();

  post.isVoted = null;

  const { container, getByLabelText, rerender } = renderComponent({
    isAuthenticated: true,
    dispatchUpVote,
    dispatchDownVote,
  }) as RenderResult;

  act(() => {
    fireEvent.click(getByLabelText('up-vote'));

    post.isVoted = PostIsVotedEnum.Up;
    post.upCount += 1;
    rerender(
      <PostVotes
        isAuthenticated
        post={post}
        dispatchUpVote={dispatchUpVote}
        dispatchDownVote={dispatchDownVote}
      />,
    );
  });

  expect(dispatchUpVote).toHaveBeenCalledTimes(1);
  expect(container.querySelector('#up-vote-count')?.textContent).toEqual(
    post.upCount.toString(),
  );
});

test('increases down voted count when down vote click', () => {
  const dispatchUpVote = jest.fn();
  const dispatchDownVote = jest.fn();

  post.isVoted = null;

  const { container, getByLabelText, rerender } = renderComponent({
    isAuthenticated: true,
    dispatchUpVote,
    dispatchDownVote,
  }) as RenderResult;

  act(() => {
    fireEvent.click(getByLabelText('down-vote'));

    post.isVoted = PostIsVotedEnum.Down;
    post.downCount -= 1;
    rerender(
      <PostVotes
        isAuthenticated
        post={post}
        dispatchUpVote={dispatchUpVote}
        dispatchDownVote={dispatchDownVote}
      />,
    );
  });

  expect(dispatchDownVote).toHaveBeenCalledTimes(1);
  expect(container.querySelector('#down-vote-count')?.textContent).toEqual(
    post.downCount.toString(),
  );
});
