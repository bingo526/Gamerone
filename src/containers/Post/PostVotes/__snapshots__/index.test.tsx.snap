// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correct number counts 1`] = `
<DocumentFragment>
  <div>
    <button
      aria-label="up-vote"
      class="button button--"
    >
      <span
        style="margin-right: 10px;"
      >
        Up
      </span>
      <span
        class="badge badge-light"
        id="up-vote-count"
      >
        5
      </span>
    </button>
    <button
      aria-label="down-vote"
      class="button button--"
    >
      <span
        style="margin-right: 10px;"
      >
        Down
      </span>
      <span
        class="badge badge-light"
        id="down-vote-count"
      >
        3
      </span>
    </button>
    <button
      class="button button--"
    >
      Comments 
      <span
        class="badge badge-light"
        id="comment-count"
      >
        0
      </span>
    </button>
  </div>
</DocumentFragment>
`;
